platform/linux-generic/odp_ipsec.c

```diff
@@ -566,17 +601,44 @@ static int ipsec_in_esp(odp_packet_t *pkt,
 	param->auth_iv_ptr = state->iv;
 
 	state->esp.aad.spi = esp.spi;
-	state->esp.aad.seq_no = esp.seq_no;
 	state->in.seq_no = odp_be_to_cpu_32(esp.seq_no);
 
+	/* AAD with ESN */
+	if (ipsec_sa->esn) {
+		state->in.seq64 = ipsec_compute_esn(ipsec_sa, state->in.seq_no);
+		if (ipsec_sa->combined_mode)
+			state->esp.aad.seq_no64 = odp_cpu_to_be_64(state->in.seq64);
+	} else {
+		state->esp.aad.seq_no = esp.seq_no;
+		state->in.seq64 = state->in.seq_no;
+	}
 	param->aad_ptr = (uint8_t *)&state->esp.aad;
 
+	/* Insert high-order bits of ESN before the ICV for ICV check
+	 * with non-combined mode algorithms.
+	 */
+	if (ipsec_sa->esn && !ipsec_sa->combined_mode) {
```

Comment:
The insert_seq_hi flag should be used here instead of this more complicated logic. And the same comment in the 4 other occurrences of the same test.

Reply:
Ack, thought about it, but didn't check this. I will check the feasibility, it seems to be possible.

Reply:
yes, everything is fine with "insert_seq_hi" flag.